File: truevfs-ext/truevfs-ext-pacemaker/src/main/scala/net/java/truevfs/ext/pacemaker/PaceModel.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.pacemaker

import net.java.truevfs.comp.inst.InstrumentingModel
import net.java.truevfs.kernel.spec.FsModel

/** A pace model.
  *
  * @author Christian Schlichtherle
  */
private final class PaceModel(mediator: PaceMediator, model: FsModel)
  extends InstrumentingModel(mediator, model) {

  private val cachedMountPoints = mediator.cachedMountPoints

  override def setMounted(isMounted: Boolean) {
    val wasMounted = model.isMounted
    model setMounted isMounted
    if (wasMounted) {
      if (!isMounted)
        cachedMountPoints remove getMountPoint
    } else {
      if (isMounted)
        cachedMountPoints put (getMountPoint, model)
    }
  }
}

File: truevfs-ext/truevfs-ext-pacemaker/src/main/java/net/java/truevfs/ext/pacemaker/PaceModel.java
/*
 * Copyright (C) 2005-2020 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.pacemaker;

import lombok.val;
import net.java.truevfs.comp.inst.InstrumentingModel;
import net.java.truevfs.kernel.spec.FsModel;
import net.java.truevfs.kernel.spec.FsMountPoint;

/**
 * A pace model.
 *
 * @author Christian Schlichtherle
 */
final class PaceModel extends InstrumentingModel<PaceMediator> {

    private final LruCache<FsMountPoint> cachedMountPoints;

    PaceModel(PaceMediator mediator, FsModel model) {
        super(mediator, model);
        this.cachedMountPoints = mediator.cachedMountPoints;
    }

    @Override
    public void setMounted(final boolean isMounted) {
        val wasMounted = model.isMounted();
        model.setMounted(isMounted);
        if (wasMounted) {
            if (!isMounted) {
                cachedMountPoints.remove(getMountPoint());
            }
        } else {
            if (isMounted) {
                cachedMountPoints.add(getMountPoint());
            }
        }
    }
}
